File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/BinaryTree/sumOfLeftLeaves_404.java
package algo.tzashinorpu.SpecialSubject.BinaryTree;

import algo.tzashinorpu.SpecialSubject.TreeNode;

public class sumOfLeftLeaves_404 {
    public int sumOfLeftLeaves_1(TreeNode root) {
        if (root == null) return 0;
        int ans = 0;
        if (root.left != null) {
            if (root.left.left == null && root.left.right == null) ans += root.left.val;
            else ans += sumOfLeftLeaves_1(root.left);
        }
        ans += sumOfLeftLeaves_1(root.right);

        return ans;
    }


    public int sumOfLeftLeaves_2(TreeNode root) {
        traverse(root);
        return sum;
    }

    // 记录左叶子之和
    int sum = 0;

    // 二叉树遍历函数
    void traverse(TreeNode root) {
        if (root == null) {
            return;
        }

        if (root.left != null &&
                root.left.left == null && root.left.right == null) {
            // 找到左侧的叶子节点，记录累加值
            sum += root.left.val;
        }

        // 递归框架
        traverse(root.left);
        traverse(root.right);
    }
}

File: Clion/ZeroTrac/Tree/sumOfLeftLeaves_404.cpp
//
// Created by TzashiNorpu on 10/31/2022.
//
#include "tree.h"

using namespace Tree;

int Solution::sumOfLeftLeaves(TreeNode *root) {
  int sum = 0;
  function<void(TreeNode *)> helper = [&](TreeNode *node) -> void {
      if (!node) return;
      if (node->left && !node->left->right && !node->left->left) sum += node->left->val;
      helper(node->left);
      helper(node->right);
  };
  helper(root);
  return sum;
}